connector/src/main/scala/org/apache/spark/sql/cassandra/DsePredicateRules.scala

```diff
@@ -53,6 +65,79 @@ object DsePredicateRules extends CassandraPredicateRules {
+  // pushdown - we might consider applying SAI later then IN rules.
+  private def saiIndexes(table: TableDef): Seq[IndexDef] = {
+    table.indexes.filter(_.className.contains(StorageAttachedIndex))
+  }
+
+  private def pushDownPredicates(predicates: AnalyzedPredicates, table: TableDef)
+    (eligibleForPushDown: Filter => Boolean): AnalyzedPredicates = {
+    val indexColumns = saiIndexes(table).map(_.targetColumn)
+    val pushedEqualityPredicates = predicates.handledByCassandra.collect {
+      case f if FilterOps.isEqualToPredicate(f) => FilterOps.columnName(f)
+    }.to[collection.mutable.Set]
+
+    val (handledByCassandra, handledBySpark) = predicates.handledBySpark.partition { filter =>
+      lazy val columnName = FilterOps.columnName(filter)
+      lazy val eligibleForSAIPushDown = eligibleForPushDown(filter) &&
+        indexColumns.contains(columnName) &&
+        !pushedEqualityPredicates.contains(columnName)
+
+      val isEqualityPredicate = FilterOps.isEqualToPredicate(filter)
+      if (isEqualityPredicate && eligibleForSAIPushDown) {
+        pushedEqualityPredicates += columnName
+      }
+      eligibleForSAIPushDown
+    }
+
+    AnalyzedPredicates(predicates.handledByCassandra ++ handledByCassandra, handledBySpark)
+  }
+
+  private def pushDownSAITextPredicates(predicates: AnalyzedPredicates, table: TableDef): AnalyzedPredicates =
+    pushDownPredicates(predicates, table) { filter =>
+      FilterOps.isEqualToPredicate(filter) &&
+        saiTextTypes.contains(table.columnByName(FilterOps.columnName(filter)).columnType)
+    }
+
+  private def pushDownSAINumericPredicates(predicates: AnalyzedPredicates, table: TableDef): AnalyzedPredicates =
+    pushDownPredicates(predicates, table) { filter =>
+      (FilterOps.isEqualToPredicate(filter) || FilterOps.isRangePredicate(filter)) &&
+        saiNumericTypes.contains(table.columnByName(FilterOps.columnName(filter)).columnType)
+    }
+
+  /**
+    * 1. IN clause for a partition key part prevents SAI push downs
+    * 2. OR clause prevents any push down
+    * 3. At most one equality is pushed
+    * 4. Only equality is pushed if equality and range are defined
+    * 5. Supported predicates:
+    *  - numeric types: >, <, >=, <=, =
+    *  - text types: =
+    *  - set and list types: contains (non-frozen collections), = (frozen collections)
+    *  - map type: contains key, contains value, key lookup result equality (map(key) == value), = (frozen collections)
+    *
+    *  NOTE that collection types support is going to be introduces in SPARKC-630
+    */
+  private def storageAttachedIndexPredicateRules(predicates: AnalyzedPredicates, table: TableDef): Seq[PredicateRule] = {
+    val partitionKeyColumns = table.partitionKey.map(_.columnName).toSet
+    val inClauseInPartitionKey = predicates.handledByCassandra
+      .find(f => FilterOps.isInPredicate(f) && f.references.exists(partitionKeyColumns.contains))
+
+    if (inClauseInPartitionKey.isDefined) {
```

Comment:
i might have misunderstood that but can we do push down sai predicates when all partition key columns are defined in `in` clause?

Reply:
no we can't, in this case we get
```
InvalidRequest: Error from server: code=2200 [Invalid query] message="Select on indexed columns and with IN clause for the PRIMARY KEY are not supported"
```